src/coreclr/vm/pgo.cpp

```diff
@@ -285,206 +322,516 @@ void PgoManager::ReadPgoData()
+    {
+        return E_NOTIMPL;
+    }
+    size_t unsafeAllocationSize = allocationSize.Value();
+    HeaderList* pHeaderList = NULL;
 
-    while (index < maxIndex)
+    if (pMD->IsDynamicMethod())
     {
-        // The first two "records" of each entry are actually header data
-        // to identify the method.
-        //
-        Header* const header = (Header*)&s_PgoData[index];
+        HeaderList *currentHeaderList = m_pgoHeaders;
+        if (currentHeaderList != NULL)
+        {
+            if (!ComparePgoSchemaEquals(currentHeaderList->header.GetData(), currentHeaderList->header.countsOffset, pSchema, countSchemaItems))
+            {
+                return E_NOTIMPL;
+            }
+            _ASSERTE(currentHeaderList->header.method == pMD);
+            *pInstrumentationData = currentHeaderList->header.GetData();
+            return S_OK;
+        }
 
-        // Sanity check that header data looks reasonable. If not, just
-        // fail the lookup.
-        //
-        if ((header->recordCount < MIN_RECORD_COUNT) || (header->recordCount > MAX_RECORD_COUNT))
+        pHeaderList = (HeaderList*)pMD->AsDynamicMethodDesc()->GetResolver()->GetJitMetaHeap()->New(unsafeAllocationSize);
+
+        memset(pHeaderList, 0, unsafeAllocationSize);
+        pHeaderList->header.Init(pMD, codehash, ilSize, offsetOfInstrumentationDataFromStartOfDataRegion);
+        *pInstrumentationData = pHeaderList->header.GetData();
+        if (!WriteInstrumentationSchema(pSchema, countSchemaItems, *pInstrumentationData, pHeaderList->header.countsOffset))
         {
-            break;
+            _ASSERTE(!"Unable to write schema");
+            return E_NOTIMPL;
         }
+        m_pgoHeaders = pHeaderList;
+        return S_OK;
+    }
+    else
+    {
+        LoaderAllocatorPgoManager *laPgoManagerThis = (LoaderAllocatorPgoManager *)this;
+        CrstHolder (&laPgoManagerThis->m_lock);
 
-        // See if the header info matches the current method.
-        //
-        if ((header->token == token) && (header->hash == hash) && (header->ilSize == ilSize))
+        HeaderList* existingData = laPgoManagerThis->m_pgoDataLookup.Lookup(pMD);
+        if (existingData != NULL)
         {
-            // Yep, found data.
-            //
-            *pBlockCounts = &s_PgoData[index + 2];
-            *pCount       = header->recordCount - 2;
-            *pNumRuns     = 1;
+            if (!ComparePgoSchemaEquals(existingData->header.GetData(), existingData->header.countsOffset, pSchema, countSchemaItems))
+            {
+                return E_NOTIMPL;
+            }
+            *pInstrumentationData = existingData->header.GetData();
             return S_OK;
         }
 
-        index += header->recordCount;
-        methodsChecked++;
+        AllocMemTracker loaderHeapAllocation;
+        pHeaderList = (HeaderList*)loaderHeapAllocation.Track(pMD->GetLoaderAllocator()->GetHighFrequencyHeap()->AllocMem(allocationSize));
```

Comment:
Seems like the schema could go in the low frequency heap but maybe that's more trouble than it's worth.

Reply:
Well, the jit interface api allows for doing things like that (which is why the Offset field is a full size_t), but for now, it would cause problems for the superpmi implementation I've come up with, and its not super clear how valuable the high vs low frequency heap actually is.